### PR TITLE
Do not print nrn_state if breakpoint does not get solve block

### PR DESCRIPTION
**Pull request by alkino**

### CODE REVIEW

**Review by iomaganaris — COMMENTED** [review]
May we add a unit test also for this to avoid any confusion?

File: src/codegen/codegen_cpp_visitor.cpp [hunk 1]
```diff
@@ -389,7 +389,7 @@ bool CodegenCVisitor::nrn_state_required() const noexcept {
     if (info.artificial_cell) {
         return false;
     }
-    return info.nrn_state_block != nullptr || breakpoint_exist();
+    return info.nrn_state_block != nullptr || info.breakpoint_with_solve || info.currents.empty();
```
Comment (pramodk) [inline src/codegen/codegen_cpp_visitor.cpp:392]: Just wanted to verify if `info.currents.empty()` is the same as `currents->next == currents` in MOD2C.

**Review by pramodk — APPROVED** [review]
LGTM

File: test/unit/codegen/codegen_cpp_visitor.cpp [hunk 2]
```diff
@@ -1159,3 +1159,46 @@ SCENARIO("Check codegen for MUTEX and PROTECT", "[codegen][mutex_protect]") {
...
+            }
+        )";
+        THEN("nrn_state should be generated") {
+            auto const generated = get_cpp_code(nmodl_text);
+            std::string expected_code =
+                R"(void nrn_state_(NrnThread* nt, Memb_list* ml, int type) {)";
+            REQUIRE_THAT(generated, Contains(expected_code));
+        }
+    }
+    GIVEN("A mod file with a SOLVE inside a BREAKPOINT") {
```
Comment (iomaganaris) [inline test/unit/codegen/codegen_cpp_visitor.cpp:1191]:
```suggestion
    GIVEN("A mod file with a STATE block") {
```

File: src/codegen/codegen_cpp_visitor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alkino) [reply src/codegen/codegen_cpp_visitor.cpp:392]: no idea, I don't understand the whole code

[the remark this answers is not in the document]
Comment (iomaganaris) [reply src/codegen/codegen_cpp_visitor.cpp:392]: Looking at the generated code I think that `NRN_STATE` blocks are basically generated by the `SOLVE` statements inside the `BREAKPOINT` blocks. Maybe @pramodk can verify this?

[the remark this answers is not in the document]
Comment (alkino) [reply src/codegen/codegen_cpp_visitor.cpp:392]: I'm not sure about currents.empty() neither

[the remark this answers is not in the document]
Comment (alkino) [reply src/codegen/codegen_cpp_visitor.cpp:392]: maybe we should drop this PR xD

[the remark this answers is not in the document]
Comment (iomaganaris) [reply src/codegen/codegen_cpp_visitor.cpp:392]: > I'm not sure about currents.empty() neither

I think this is actually needed. The only difference in the tests I suggested I see with `mod2c` is just that `mod2c` registers the `nrn_state_` function in the `register_mech()` function, while `NMODL` does not. I think this is an improvement though, so it should be okay

**Comment by iomaganaris** [comment]
With Nico we discussed that since this PR controls something pretty important it's better not to merge at this state, before adding sufficient tests and validating that `NMODL` has the same bahavior as `mod2c`

**Comment by 1uc** [comment]
@alkino can this be closed?

**Pull request closed without merging**